Формат CSV

CSV (от англ. Comma-Separated Values — значения, разделённые запятыми) — текстовый формат, предназначенный для представления табличных данных. Строка таблицы соответствует строке текста, которая содержит одно или несколько полей, разделенных запятыми.

Например, таблица:

    Rank   	       Language     	        Share       
1	Python	31.17%
2	Java	17.75%
3	JavaScript	8%
4	C#	7.05%
5	PHP	6.09%
в формате csv будет выглядеть так:

Rank,Language,Share
1,Python,31.17%
2,Java,17.75%
3,JavaScript,8%
4,C#,7.05%
5,PHP,6.09%

Обратите внимание, пробелов после запятой быть не должно.

Ручная работа с файлами

Рассмотрим текстовый файл products.csv, содержащий информацию о товарах некоторого интернет магазина. Файл содержит информацию о трех столбцах:

keywords (ключевые слова)
price (цена)
product_name (имя продукта)
и имеет следующее содержимое:

keywords,price,product_name
Садовый стул,1699,ВЭДДО
Садовый стул,2999,ЭПЛАРО
Садовый табурет,1699,ЭПЛАРО
Садовый стол,1999,ТЭРНО
Складной стол,7499,ЭПЛАРО
Настил,1299,РУННЕН
Стеллаж,1299,ХИЛЛИС
Кружка,39,СТЕЛЬНА
Молочник,299,ВАРДАГЕН
Термос для еды,699,ЭФТЕРФРОГАД
Ситечко,59,ИДЕАЛИСК
Чайник заварочный,499,РИКЛИГ
Кофе-пресс,699,УПХЕТТА
Чашка с блюдцем,249,ИКЕА
Кружка,249,ЭМНТ
Ситечко,199,САККУННИГ
Кружка,199,ФИНСТИЛТ
Тарелка,269,ЭВЕРЕНС

Разделителем записей был выбран символ новой строки, а разделителем полей — символ запятой.

В результате мы получили текстовый файл, который легко читается и человеком и компьютерной программой.

Поскольку любой csv файл является текстовым, то нам не составит труда обработать его "руками", подобно тому, как мы обрабатываем обычный текстовый файл.

In [6]:
with open('products.csv', encoding='utf-8') as file:
    data = file.read()
    for line in data.splitlines():
        print(line.split(','))

['keywords', 'price', 'product_name']
['Садовый стул', '1699', 'ВЭДДО']
['Садовый стул', '2999', 'ЭПЛАРО']
['Садовый табурет', '1699', 'ЭПЛАРО']
['Садовый стол', '1999', 'ТЭРНО']
['Складной стол', '7499', 'ЭПЛАРО']
['Настил', '1299', 'РУННЕН']
['Стеллаж', '1299', 'ХИЛЛИС']
['Кружка', '39', 'СТЕЛЬНА']
['Молочник', '299', 'ВАРДАГЕН']
['Термос для еды', '699', 'ЭФТЕРФРОГАД']
['Ситечко', '59', 'ИДЕАЛИСК']
['Чайник заварочный', '499', 'РИКЛИГ']
['Кофе-пресс', '699', 'УПХЕТТА']
['Чашка с блюдцем', '249', 'ИКЕА']
['Кружка', '249', 'ЭМНТ']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Тарелка', '269', 'ЭВЕРЕНС']


Для построчного разделения текста удобно использовать строковый метод splitlines(), вместо метода split('\n')

Мы также можем создать вложенный список (таблицу) для более удобного взаимодействия с данными:

In [12]:
with open('products.csv', encoding='utf-8') as file:
    data = file.read()
    table = [r.split(',') for r in data.splitlines()]
    # print(*table)
    for line in table:
        print(line)

['keywords', 'price', 'product_name']
['Садовый стул', '1699', 'ВЭДДО']
['Садовый стул', '2999', 'ЭПЛАРО']
['Садовый табурет', '1699', 'ЭПЛАРО']
['Садовый стол', '1999', 'ТЭРНО']
['Складной стол', '7499', 'ЭПЛАРО']
['Настил', '1299', 'РУННЕН']
['Стеллаж', '1299', 'ХИЛЛИС']
['Кружка', '39', 'СТЕЛЬНА']
['Молочник', '299', 'ВАРДАГЕН']
['Термос для еды', '699', 'ЭФТЕРФРОГАД']
['Ситечко', '59', 'ИДЕАЛИСК']
['Чайник заварочный', '499', 'РИКЛИГ']
['Кофе-пресс', '699', 'УПХЕТТА']
['Чашка с блюдцем', '249', 'ИКЕА']
['Кружка', '249', 'ЭМНТ']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Тарелка', '269', 'ЭВЕРЕНС']


С помощью такой таблицы мы можем обратиться к цене седьмого по счету товара:

In [10]:
print(table[7][1])

1299


Или мы можем также отсортировать товары по цене и напечатать 5 самых дешевых товаров.

In [11]:
with open('products.csv', encoding='utf-8') as file:
    data = file.read()
    table = [r.split(',') for r in data.splitlines()]
    del table[0]                                        # удаляем заголовок
    table.sort(key=lambda item: int(item[1]))
    for line in table[:5]:
        print(line)

['Кружка', '39', 'СТЕЛЬНА']
['Ситечко', '59', 'ИДЕАЛИСК']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Чашка с блюдцем', '249', 'ИКЕА']


Обратите внимание на то, что мы удалили первый элемент из списка, так как он содержит не данные, а заголовки. Также стоит обратить внимание на то, что в лямбда-функции мы преобразуем элемент item[1] к числовому типу, иначе сортировка будет работать не так, как полагается.

В приведенном выше файле products.csv символ запятой использовался в качестве разделителя полей. Однако бывают ситуации, в которых сам символ запятой является легитимным символом. Например, столбец keywords может содержать (и как правило на практике содержит) несколько ключевых слов. В таком случае структура файла нарушается. Для решения такой проблемы можно использовать два подхода.

1 подход. Если поле содержит запятые, то это поле должно быть заключено в двойные кавычки. Если этого не сделать, то данные невозможно будет корректно обработать.

Содержимое файла products.csv может иметь вид:

keywords,price,product_name
"Садовый стул, стул для дачи",1699,ВЭДДО
Садовый стул,2999,ЭПЛАРО
Садовый табурет,1699,ЭПЛАРО
Садовый стол,1999,ТЭРНО
"Складной стол, обеденный стол",7499,ЭПЛАРО
Настил,1299,РУННЕН
Стеллаж,1299,ХИЛЛИС
"Кружка, сосуд, стакан с ручкой",39,СТЕЛЬНА
Молочник,299,ВАРДАГЕН
Термос для еды,699,ЭФТЕРФРОГАД
Ситечко,59,ИДЕАЛИСК
Чайник заварочный,499,РИКЛИГ
Кофе-пресс,699,УПХЕТТА
Чашка с блюдцем,249,ИКЕА
"Кружка, стакан с ручкой",249,ЭМНТ
Ситечко,199,САККУННИГ
Кружка,199,ФИНСТИЛТ
"Тарелка, блюдце",269,ЭВЕРЕНС

Обратите внимание на записи с номерами 1, 5, 8, 15 и 18. В этих записях в столбце keywords используется символ запятой, и для правильной обработки данных нам необходимо соответствующее поле обрамить символом двойных кавычек.

Обратите внимание на то, что при таком содержимом файла products.csv наш код, написанный выше, является нерабочим, так как он разделит каждую строку через символ запятой.

2 подход. Использовать в качестве разделителя другой символ, например, символ табуляции (\t), который весьма редко встречается в качестве валидного содержимого файла.

keywords	price	product_name
Садовый стул, стул для дачи	1699	ВЭДДО
Садовый стул	2999	ЭПЛАРО
Садовый табурет	1699	ЭПЛАРО
Садовый стол	1999	ТЭРНО
Складной стол, обеденный стол	7499	ЭПЛАРО
Настил	1299	РУННЕН
Стеллаж	1299	ХИЛЛИС
Кружка, сосуд, стакан с ручкой	39	СТЕЛЬНА
Молочник	299	ВАРДАГЕН
Термос для еды	699	ЭФТЕРФРОГАД
Ситечко	59	ИДЕАЛИСК
Чайник заварочный	499	РИКЛИГ
Кофе-пресс	699	УПХЕТТА
Чашка с блюдцем	249	ИКЕА
Кружка, стакан с ручкой	249	ЭМНТ
Ситечко	199	САККУННИГ
Кружка	199	ФИНСТИЛТ
Тарелка, блюдце	269	ЭВЕРЕНС

Формат данных csv, в котором разделителем является символ табуляции, называют tsv (англ. tab separated values — «значения, разделенные табуляцией»).

tsv — текстовый формат для представления табличных данных. Каждая запись в таблице — строка текстового файла. Каждое поле записи отделяется от других символом табуляции, а точнее — горизонтальной табуляции.

tsv и csv — формы более общего формата dsv (англ. delimiter separated values — «значения, разграниченные разделителем»).

Приведенный ниже код правильно обрабатывает файл products.tsv, в котором символом разделителя выбран символ табуляции (\t):

In [13]:
with open('products.tsv', encoding='utf-8') as file:
    data = file.read()
    table = [r.split('\t') for r in data.splitlines()]
    del table[0]
    table.sort(key=lambda item: int(item[1]))
    for line in table[:5]:
        print(line)

['Кружка, сосуд, стакан с ручкой', '39', 'СТЕЛЬНА']
['Ситечко', '59', 'ИДЕАЛИСК']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Чашка с блюдцем', '249', 'ИКЕА']


Модуль csv

Несмотря на то что csv формат очень прост и мы можем работать с ним, как с обычным текстовым файлом, на практике используется встроенный модуль csv.

В данном модуле есть два основных объекта: reader и writer, созданные, чтобы читать и создавать csv файлы соответственно.

Чтение данных с помощью reader

Для импорта модуля мы используем строку кода:

In [14]:
import csv

Этот модуль входит в стандартную библиотеку, и его не нужно устанавливать каким‑то особенным способом.

Рассмотрим все тот же файл products.csv, содержащий информацию о товарах интернет магазина:

keywords,price,product_name
Садовый стул,1699,ВЭДДО
Садовый стул,2999,ЭПЛАРО
Садовый табурет,1699,ЭПЛАРО
Садовый стол,1999,ТЭРНО
Складной стол,7499,ЭПЛАРО
Настил,1299,РУННЕН
Стеллаж,1299,ХИЛЛИС
Кружка,39,СТЕЛЬНА
Молочник,299,ВАРДАГЕН
Термос для еды,699,ЭФТЕРФРОГАД
Ситечко,59,ИДЕАЛИСК
Чайник заварочный,499,РИКЛИГ
Кофе-пресс,699,УПХЕТТА
Чашка с блюдцем,249,ИКЕА
Кружка,249,ЭМНТ
Ситечко,199,САККУННИГ
Кружка,199,ФИНСТИЛТ
Тарелка,269,ЭВЕРЕНС

In [15]:
import csv

with open('products.csv', encoding='utf-8') as file:
    rows = csv.reader(file)                               # создаем reader объект
    for row in rows:
        print(row)

['keywords', 'price', 'product_name']
['Садовый стул', '1699', 'ВЭДДО']
['Садовый стул', '2999', 'ЭПЛАРО']
['Садовый табурет', '1699', 'ЭПЛАРО']
['Садовый стол', '1999', 'ТЭРНО']
['Складной стол', '7499', 'ЭПЛАРО']
['Настил', '1299', 'РУННЕН']
['Стеллаж', '1299', 'ХИЛЛИС']
['Кружка', '39', 'СТЕЛЬНА']
['Молочник', '299', 'ВАРДАГЕН']
['Термос для еды', '699', 'ЭФТЕРФРОГАД']
['Ситечко', '59', 'ИДЕАЛИСК']
['Чайник заварочный', '499', 'РИКЛИГ']
['Кофе-пресс', '699', 'УПХЕТТА']
['Чашка с блюдцем', '249', 'ИКЕА']
['Кружка', '249', 'ЭМНТ']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Тарелка', '269', 'ЭВЕРЕНС']


Самая важная строка кода в программе — это строка с созданием reader объекта:

In [ ]:
rows = csv.reader(file)

Объект reader дает доступ к построчному итератору, полностью аналогичному работе с файлом или списком.

После выполнения этой строки в переменную rows будет записан итератор, с помощью которого можно «пробежаться» циклом по файлу. В каждой итерации цикла при этом будет доступна соответствующая строка файла, уже разбитая по запятым и представляющая собой список. При этом автоматически будут учтены все нюансы с запятыми внутри кавычек и самими кавычками.

In [38]:
import csv

with open('products1.csv', encoding='utf-8') as file:
    rows = csv.reader(file)
    # for row in rows:
    #     print(row)
    # table = [row for row in rows]
    table = list(rows)
    for row in table:
        print(row)

['keywords', 'price', 'product_name']
['Садовый стул, стул для дачи', '1699', 'ВЭДДО']
['Садовый стул', '2999', 'ЭПЛАРО']
['Садовый табурет', '1699', 'ЭПЛАРО']
['Садовый стол', '1999', 'ТЭРНО']
['Складной стол, обеденный стол', '7499', 'ЭПЛАРО']
['Настил', '1299', 'РУННЕН']
['Стеллаж', '1299', 'ХИЛЛИС']
['Кружка, сосуд, стакан с ручкой', '39', 'СТЕЛЬНА']
['Молочник', '299', 'ВАРДАГЕН']
['Термос для еды', '699', 'ЭФТЕРФРОГАД']
['Ситечко', '59', 'ИДЕАЛИСК']
['Чайник заварочный', '499', 'РИКЛИГ']
['Кофе-пресс', '699', 'УПХЕТТА']
['Чашка с блюдцем', '249', 'ИКЕА']
['Кружка, стакан с ручкой', '249', 'ЭМНТ']
['Ситечко', '199', 'САККУННИГ']
['Кружка', '199', 'ФИНСТИЛТ']
['Тарелка, блюдце', '269', 'ЭВЕРЕНС']


In [39]:
print(table[1][0])
print(table[2][0])

Садовый стул, стул для дачи
Садовый стул


Так как каждая строка файла, полученная из итератора, является списком, к ней можно применять все способы работы со списками.

In [40]:
import csv

with open('products.csv', encoding='utf-8') as file:
    rows = csv.reader(file)
    for keywords, price, product_name in rows:
        print(f'Ключевые слова: {keywords}, цена: {price}, название: {product_name}')

Ключевые слова: keywords, цена: price, название: product_name
Ключевые слова: Садовый стул, цена: 1699, название: ВЭДДО
Ключевые слова: Садовый стул, цена: 2999, название: ЭПЛАРО
Ключевые слова: Садовый табурет, цена: 1699, название: ЭПЛАРО
Ключевые слова: Садовый стол, цена: 1999, название: ТЭРНО
Ключевые слова: Складной стол, цена: 7499, название: ЭПЛАРО
Ключевые слова: Настил, цена: 1299, название: РУННЕН
Ключевые слова: Стеллаж, цена: 1299, название: ХИЛЛИС
Ключевые слова: Кружка, цена: 39, название: СТЕЛЬНА
Ключевые слова: Молочник, цена: 299, название: ВАРДАГЕН
Ключевые слова: Термос для еды, цена: 699, название: ЭФТЕРФРОГАД
Ключевые слова: Ситечко, цена: 59, название: ИДЕАЛИСК
Ключевые слова: Чайник заварочный, цена: 499, название: РИКЛИГ
Ключевые слова: Кофе-пресс, цена: 699, название: УПХЕТТА
Ключевые слова: Чашка с блюдцем, цена: 249, название: ИКЕА
Ключевые слова: Кружка, цена: 249, название: ЭМНТ
Ключевые слова: Ситечко, цена: 199, название: САККУННИГ
Ключевые слова: Кружка

При создании reader объекта мы можем его настраивать, указывая:

аргумент delimiter — односимвольная строка, используемая для разделения полей, по умолчанию имеет значение ','
аргумент quotechar — односимвольная строка, используемая для кавычек в полях, содержащих специальные символы, по умолчанию имеет значение '"'.

Пусть содержимое файла products.csv имеет вид (в качестве разделителя выбран символ ';'):

keywords;price;product_name
"Садовый стул, стул для дачи";1699;ВЭДДО
Садовый стул;2999;ЭПЛАРО
Садовый табурет;1699;ЭПЛАРО
Садовый стол;1999;ТЭРНО
"Складной стол, обеденный стол";7499;ЭПЛАРО
Настил;1299;РУННЕН
Стеллаж;1299;ХИЛЛИС
"Кружка, сосуд, стакан с ручкой";39;СТЕЛЬНА
Молочник;299;ВАРДАГЕН
Термос для еды;699;ЭФТЕРФРОГАД
Ситечко;59;ИДЕАЛИСК
Чайник заварочный;499;РИКЛИГ
Кофе-пресс;699;УПХЕТТА
Чашка с блюдцем;249;ИКЕА
"Кружка, стакан с ручкой";249;ЭМНТ
Ситечко;199;САККУННИГ
Кружка;199;ФИНСТИЛТ
"Тарелка, блюдце";269;ЭВЕРЕНС

In [41]:
import csv

with open('products2.csv', encoding='utf-8') as file:
    rows = csv.reader(file, delimiter=';', quotechar='"')
    for index, row in enumerate(rows):
        if index > 5:
            break
        print(row)

['keywords', 'price', 'product_name']
['Садовый стул, стул для дачи', '1699', 'ВЭДДО']
['Садовый стул', '2999', 'ЭПЛАРО']
['Садовый табурет', '1699', 'ЭПЛАРО']
['Садовый стол', '1999', 'ТЭРНО']
['Складной стол, обеденный стол', '7499', 'ЭПЛАРО']


При создании reader объекта мы указываем, что символ-разделитель записей delimiter в нашем файле — точка с запятой, а символ кавычек quotechar — двойные кавычки. Кроме того, мы используем встроенную функцию enumerate() для нумерации строк.

Обратите внимание на то, что для корректной обработки данных мы должны все еще исключить первую строку из обработки, которая содержит названия столбцов.

Чтение данных с помощью DictReader

Иcпользовать reader объект не всегда удобно, так как он возвращает сырые списки из строк файла, к тому же первой строкой является строка с названиями столбцов, которая практически всегда удаляется, так как мешает правильной обработке данных.

В модуле csv есть специальный объект DictReader, который поддерживает создание объекта-словаря на основе названий столбцов. С помощью DictReader объекта мы можем обращаться к полям не по индексу, а по названию, что делает код более понятным.

In [43]:
import csv

with open('products3.csv', encoding='utf-8') as file:
    rows = csv.DictReader(file, delimiter=';', quotechar='"')
    for row in rows:
        print(row)

{'keywords': 'Садовый стул, стул для дачи', 'price': '1699', 'product_name': 'ВЭДДО'}
{'keywords': 'Садовый стул', 'price': '2999', 'product_name': 'ЭПЛАРО'}
{'keywords': 'Садовый табурет', 'price': '1699', 'product_name': 'ЭПЛАРО'}
{'keywords': 'Садовый стол', 'price': '1999', 'product_name': 'ТЭРНО'}
{'keywords': 'Складной стол, обеденный стол', 'price': '7499', 'product_name': 'ЭПЛАРО'}
{'keywords': 'Настил', 'price': '1299', 'product_name': 'РУННЕН'}
{'keywords': 'Стеллаж', 'price': '1299', 'product_name': 'ХИЛЛИС'}
{'keywords': 'Кружка, сосуд, стакан с ручкой', 'price': '39', 'product_name': 'СТЕЛЬНА'}
{'keywords': 'Молочник', 'price': '299', 'product_name': 'ВАРДАГЕН'}
{'keywords': 'Термос для еды', 'price': '699', 'product_name': 'ЭФТЕРФРОГАД'}
{'keywords': 'Ситечко', 'price': '59', 'product_name': 'ИДЕАЛИСК'}
{'keywords': 'Чайник заварочный', 'price': '499', 'product_name': 'РИКЛИГ'}
{'keywords': 'Кофе-пресс', 'price': '699', 'product_name': 'УПХЕТТА'}
{'keywords': 'Чашка с блю

Обратите внимание на то, что переменная row имеет тип dict в Python 3.8+. В более ранних версиях переменная row имела тип OrderedDict, который мы изучим чуть позже.

In [44]:
import csv

with open('products3.csv', encoding='utf-8') as file:
    rows = csv.DictReader(file, delimiter=';', quotechar='"')
    expensive = sorted(rows, key=lambda item: int(item['price']), reverse=True)
    for record in expensive[:5]:
        print(record)

{'keywords': 'Складной стол, обеденный стол', 'price': '7499', 'product_name': 'ЭПЛАРО'}
{'keywords': 'Садовый стул', 'price': '2999', 'product_name': 'ЭПЛАРО'}
{'keywords': 'Садовый стол', 'price': '1999', 'product_name': 'ТЭРНО'}
{'keywords': 'Садовый стул, стул для дачи', 'price': '1699', 'product_name': 'ВЭДДО'}
{'keywords': 'Садовый табурет', 'price': '1699', 'product_name': 'ЭПЛАРО'}


При создании DictReader объекта значениями по умолчанию для аргументов delimiter и quotechar являются ',' (символ запятой) и '"' (символ двойной кавычки) соответственно.

Обратите внимание на то, что при использовании DictReader мы не избавляемся от первой строки, содержащей названия столбцов. Они хранятся в атрибуте fieldnames объекта DictReader. При этом к элементам строк мы обращаемся теперь не по индексам (int(item[1])), а по их названиям (int(item['price'])), что намного удобнее.

Запись данных с помощью writer

Для записи данных в csv файл можно использовать специальный writer объект.

In [45]:
import csv

columns = ['first_name', 'second_name', 'class_number', 'class_letter']
data = [['Тимур', 'Гуев', 11, 'А'], ['Руслан', 'Чаниев', 9, 'Б'], ['Артур', 'Харисов', 10, 'В']]

with open('students.csv', 'w', encoding='utf-8', newline='') as file:
    writer = csv.writer(file)
    writer.writerow(columns)                 # запись заголовков
    for row in data:                         # запись строк
        writer.writerow(row)

создает файл students.csv с содержимым:

first_name,second_name,class_number,class_letter
Тимур,Гуев,11,А
Руслан,Чаниев,9,Б
Артур,Харисов,10,В

Обратите внимание на необязательный параметр newline функции open(), который имеет значение '' (пустой строки). Он отвечает за переводы строк при чтении или записи в текстовый файл. По умолчанию имеет значение None, в этом случае все разделители строк преобразуются в '\n'. Если в файле оказывается лишний перевод строки, то следует использовать этот параметр в режиме newline='', тогда '\n' будет преобразован в пустую строку.

При создании writer объекта мы так же можем его настраивать, задавая delimiter и многие другие параметры.

In [ ]:
import csv

columns = ['first_name', 'second_name', 'class_number', 'class_letter']
data = [['Тимур', 'Гуев', 11, 'А'], ['Руслан', 'Чаниев', 9, 'Б'], ['Роман', 'Белых', 10, 'В']]

with open('students.csv', 'w', encoding='utf-8', newline='') as file:
    writer = csv.writer(file, delimiter=';', quoting=csv.QUOTE_NONNUMERIC)
    writer.writerow(columns)
    for row in data:
        writer.writerow(row)

создает файл с содержимым:

"first_name";"second_name";"class_number";"class_letter"
"Тимур";"Гуев";11;"А"
"Руслан";"Чаниев";9;"Б"
"Роман";"Белых";10;"В"

Значение аргумента quoting=csv.QUOTE_NONNUMERIC означает, что в кавычки будут браться все нечисловые значения. По умолчанию символом кавычки является ", если нужно поменять символ, то используйте уже знакомый нам именованный аргумент quotechar

Для задания параметра quoting используются специальные константы из модуля csv:

QUOTE_ALL: указывает объектам записи указывать все поля
QUOTE_MINIMAL: указывает объектам записи заключать в кавычки только те поля, которые содержат специальные символы, такие как разделитель delimiter, кавычка quotechar или любой из символов в lineterminator
QUOTE_NONNUMERIC: указывает объектам записи указывать все нечисловые поля
QUOTE_NONE: указывает объектам записи никогда не заключать в кавычки поля

Помимо метода writerow() можно использовать и метод writerows(), чтобы записать сразу несколько строк. Единственным аргументом этого метода может быть коллекция коллекций. То есть, каждый элемент списка rows в нашем случае должен быть коллекцией. Если rows будет, например, списком чисел, программа завершится с ошибкой.

In [ ]:
import csv

columns = ['first_name', 'second_name', 'class_number', 'class_letter']
data = [['Тимур', 'Гуев', 11, 'А'], ['Руслан', 'Чаниев', 9, 'Б'], ['Роман', 'Белых', 10, 'В']]

with open('students.csv', 'w', encoding='utf-8', newline='') as file:
    writer = csv.writer(file, delimiter=';', quoting=csv.QUOTE_NONNUMERIC)
    writer.writerow(columns)
    writer.writerows(data)

создает файл students.csv с содержимым:

"first_name";"second_name";"class_number";"class_letter"
"Тимур";"Гуев";11;"А"
"Руслан";"Чаниев";9;"Б"
"Роман";"Белых";10;"В"

Запись данных с помощью DictWriter

Для записи данных в csv файл также можно использовать DictWriter объект, который позволяет записывать содержимое словаря в файл.

In [ ]:
import csv

data = [{'first_name': 'Тимур', 'second_name': 'Гуев', 'class_number': 11, 'class_letter': 'А'},
        {'first_name': 'Руслан', 'second_name': 'Чаниев', 'class_number': 9, 'class_letter': 'Б'},
        {'first_name': 'Роман', 'second_name': 'Белых', 'class_number': 10, 'class_letter': 'В'}]

columns = ['first_name', 'second_name', 'class_number', 'class_letter']

with open('students.csv', 'w', encoding='utf-8', newline='') as file:
    writer = csv.DictWriter(file, fieldnames=columns, delimiter=';', quoting=csv.QUOTE_NONNUMERIC)
    writer.writeheader()                 # запись заголовков
    for row in data:                     # запись строк
        writer.writerow(row)

создает файл students.csv с содержимым:

"first_name";"second_name";"class_number";"class_letter"
"Тимур";"Гуев";11;"А"
"Руслан";"Чаниев";9;"Б"
"Роман";"Белых";10;"В"

Мы также можем использовать метод writerows() объекта DictWriter для записи сразу нескольких строк. Таким образом вместо строк кода:

In [ ]:
for row in data:
    writer.writerow(row)

можно написать:

In [ ]:
writer.writerows(data)

Обратите внимание на то, что ключи словарей, которые записываются в файл, должны совпадать с названиями полей, которые переданы в качестве аргумента fieldnames, иначе будет возникать ошибка ValueError.

Примечания

Примечание 1. Спецификация формата csv:

каждая строка файла — это одна строка таблицы
разделителем (delimiter) значений колонок является символ запятой ,. Однако на практике часто используются другие разделители, например: ; или символ табуляции \t (формат TSV)
значения, содержащие зарезервированные символы (двойная кавычка, запятая, точка с запятой, новая строка), обрамляются двойными кавычками ". Если в значении встречаются кавычки — они представляются в файле в виде двух кавычек подряд

Примечание 4. Достаточно распространенная ошибка при чтении данных из csv файла — это передача в csv.reader (csv.DictReader) не файлового объекта, а текстовых данных, считанных из файла. Это не вызывает ошибки, но результат будет отличаться от ожидаемого.

Примечание 5. При создании reader объекта важно, чтобы файл был открыт в режиме чтения ('r'), аналогично при создании writer объекта — в режиме записи ('w').

Примечание 6. Так как reader объект является итератором, мы можем преобразовать его в список или кортеж с помощью встроенных функций list() и tuple() соответственно, но не можем преобразовать в множество, т.к. его элементами могут быть только неизменяемые объекты, которыми списки не являются.

метод  pandas.read_csv  принимает .csv файл и возвращает pandas.DataFrame (это что-то наподобие  экселевской   таблички). Если указать неверный разделитель, ошибки не возникнет -  просто получим винегрет с одной колонкой. Атрибут pandas.shape содержит кортеж из количества строк и столбцов. Идея в том, что  перебирая разделители находим корректный  DataFrame с количеством столбцов  больше одного и добавляем его в список датафреймов.

In [50]:
import pandas as pd

# Функция, которая получает список файлов и возвращает список DataFrames
def get_list_df(list_src_topics: list) -> list:
    list_df = []  # Инициализируем пустой список для хранения DataFrames
    for file in list_src_topics:  # Итерируемся по каждому файлу в переданном списке
        for s in [';', ',', '|']:  # Перебираем возможные разделители
            # Пытаемся считать файл, используя текущий разделитель
            df = pd.read_csv(file, encoding='utf-8-sig', sep=s)
            # Проверяем, сформировался ли DataFrame с более чем одной колонкой
            if df.shape[1] > 1:  # Если DataFrame имеет более одной колонки
                list_df.append(df)  # Добавляем DataFrame в список
                break  # Выходим из внутреннего цикла, если успешное считывание
    return list_df  # Возвращаем список DataFrames

# Пример вызова функции с файлом 'products.csv'
get_list_df(['products.csv'])

[             keywords  price product_name
 0        Садовый стул   1699        ВЭДДО
 1        Садовый стул   2999       ЭПЛАРО
 2     Садовый табурет   1699       ЭПЛАРО
 3        Садовый стол   1999        ТЭРНО
 4       Складной стол   7499       ЭПЛАРО
 5              Настил   1299       РУННЕН
 6             Стеллаж   1299       ХИЛЛИС
 7              Кружка     39      СТЕЛЬНА
 8            Молочник    299     ВАРДАГЕН
 9      Термос для еды    699  ЭФТЕРФРОГАД
 10            Ситечко     59     ИДЕАЛИСК
 11  Чайник заварочный    499       РИКЛИГ
 12         Кофе-пресс    699      УПХЕТТА
 13    Чашка с блюдцем    249         ИКЕА
 14             Кружка    249         ЭМНТ
 15            Ситечко    199    САККУННИГ
 16             Кружка    199     ФИНСТИЛТ
 17            Тарелка    269      ЭВЕРЕНС]

Вам доступен csv файл grades.csv, имеющий следующее содержание:

name;grade
Timur;100
Ruslan;97

Ниже представлена программа, которая должна открывать данный файл и выводить содержимое каждой строки в виде списка. В программе допущена ошибка, поэтому она выводит:

['n']
['a']
['m']
['e']
['', '']
['g']
['r']
['a']
['d']
['e']
[]
['T']
...

Найдите и исправьте ее, чтобы результатом работы программы были строки:

['name', 'grade']
['Timur', '100']
['Ruslan', '97']

In [57]:
import csv

with open('grades.csv', encoding='utf-8') as csv_file:
    # считываем содержимое файла
    # text = csv_file.read()
    # создаем reader объект и указываем в качестве разделителя символ ;
    # rows = csv.reader(text, delimiter=';')
    rows = csv.reader(csv_file, delimiter=';')
    # выводим каждую строку
    for row in rows:
        print(row)

['name', 'grade']
['Timur', '100']
['Ruslan', '97']


При попытке выполнить приведенную ниже программу возникает ошибка. Найдите и исправьте ее, чтобы программа создала файл writing_test.csv, имеющий следующее содержание:

first_col,second_col
value1,value2

In [62]:
import csv

with open('writing_test.csv', 'w', encoding='utf-8', newline='') as csv_file:
    # создаем writer объект и указываем названия столбцов
    writer = csv.DictWriter(csv_file, fieldnames=['first_col', 'second_col'])
    # записываем первую строку с названиями столбцов
    writer.writeheader()
    # записываем строку с данными
    writer.writerows([{'first_col': 'value1', 'second_col': 'value2'}])

In [ ]:
import csv

with open('writing_test.csv', 'w', encoding='utf-8') as csv_file:
    keys, data = ['first_col', 'second_col'], ['value1', 'value2']
    writer = csv.DictWriter(csv_file, fieldnames=keys)
    writer.writeheader()
    writer.writerow(dict(zip(keys, data)))

Скидки
Наступил ноябрь, и во многих магазинах начались распродажи, но как многим известно, зачастую товары со скидкой оказываются дороже, чем без нее. Вам доступен файл sales.csv, который содержит данные о ценообразовании различной бытовой техники. В первом столбце записано название товара, во втором — старая цена, в третьем — новая цена со скидкой:

name;old_price;new_price
Встраиваемая посудомоечная машина De'Longhi DDW 06S;23089;31862
Вытяжка Falmec Afrodite 60/600;27694;18001
...
Напишите программу, которая выводит названия тех товаров, цена на которые уменьшилась. Товары должны быть расположены в своем исходном порядке, каждый на отдельной строке.

Примечание 1. Разделителем в файле sales.csv является точка с запятой, при этом кавычки не используются.

Примечание 3. Начальная часть ответа выглядит так:

Вытяжка Falmec Afrodite 60/600
Духовой шкаф AEG BS 5836600
Вытяжка MAUNFELD PLYM 60
...

Примечание 4. При открытии файла используйте явное указание кодировки UTF-8.

In [106]:
import csv

with open('sales.csv', encoding='utf-8-sig') as file:
    rows = csv.DictReader(file, delimiter=';')
    # best = sorted(rows, key=lambda x:int(x['old_price'])-int(x['new_price'])>0, reverse=True)
    for row in rows:
        if int(row['old_price']) > int(row['new_price']): 
            print(row['name'])

Вытяжка Falmec Afrodite 60/600
Духовой шкаф AEG BS 5836600
Вытяжка MAUNFELD PLYM 60
Весы Momert 7462
Кухонная мойка Blanco Classic Pro 5 S-IF
Вакуумный упаковщик Hurakan HKN-V300
Морозильная камера Frostor F2000B
Холодильник Atlant XM-4210
Стиральная машина Bosch WLK 20263
Швейная машина, оверлок Toyota JET B 224
Весы Mystery MES-1805
Винный шкаф Caso WineMaster 38
Варочная поверхность Gorenje G6 N40
Вытяжка Kuppersberg Onda 50
Весы Energy ENM-408
Фен Centek CT-2234
Встраиваемый холодильник Liebherr ICS 3013
Массажер для тела Yamaguchi Axiom Matrix
Варочная поверхность Neff N 74TD00N0
Духовой шкаф Smeg SF99
Вытяжка Korting KHC 6956
Варочная поверхность Smeg SR775OT
Вытяжка Cata TF 5250
Стиральная машина Miele WMV 960 WPS
Электрочайник Bugatti VERA
Стиральная машина Bosch WAW 24440
Тостер Velikie Reki Pyshka-8
Пылесос Karcher CV 38/2
Вакуумный упаковщик Rommelsbacher VAC 155
Швейная машина, оверлок Jaguar EL-109
Встраиваемая посудомоечная машина Electrolux ESL 8820
Стиральная машина Whi

In [102]:
import csv

with open('sales.csv', encoding='utf-8-sig') as file:
    data = file.read()
    table = [i.split(';') for i in data.splitlines()]
    del table[0]
    table.sort(key=lambda x: int(x[1]) - int(x[2]) > 0, reverse=True)
    for row in table[:5]:
        print(row[0])

Вытяжка Falmec Afrodite 60/600
Духовой шкаф AEG BS 5836600
Вытяжка MAUNFELD PLYM 60
Весы Momert 7462
Кухонная мойка Blanco Classic Pro 5 S-IF


Средняя зарплата
Вам доступен файл salary_data.csv, который содержит анонимную информацию о зарплатах сотрудников в различных компаниях. В первом столбце записано название компании, а во втором — зарплата очередного сотрудника:

company_name;salary
Atos;135000
ХайТэк;24400
Philax;128600
Инлайн Груп;43900
IBS;70600
Oracle;131600
Atos;91000
...
Напишите программу, которая упорядочивает компании по возрастанию средней зарплаты ее сотрудников и выводит их названия, каждое на отдельной строке. Если две компании имеют одинаковые средние зарплаты, они должны быть расположены в лексикографическом порядке их названий.

Примечание 1. Средняя зарплата компании определяется как отношение суммы всех зарплат к их количеству.

Примечание 2. Разделителем в файле salary_data.csv является точка с запятой, при этом кавычки не используются.

Примечание 4. Начальная часть ответа выглядит так:

Информзащита
Форс
OFT group
...
Примечание 5. При открытии файла используйте явное указание кодировки UTF-8.

In [129]:
import csv

best = {}
with open('salary_data.csv', encoding='utf-8-sig') as file:
    rows = csv.DictReader(file, delimiter=';', quotechar='"')
    for row in rows:
        company = row['company_name']
        salary = int(row['salary'])
        
        if company in best:
            best[company]['total_salary'] += salary
            best[company]['count'] += 1
        else:
            best[company] = {'total_salary': salary, 'count': 1}
            
very_best = sorted(best.items(), key=lambda x:x[1]['total_salary']/x[1]['count'])

for i in very_best:
    print(i)

('Информзащита', {'total_salary': 25626900, 'count': 333})
('Форс', {'total_salary': 25669500, 'count': 333})
('OFT group', {'total_salary': 28535000, 'count': 367})
('Oracle', {'total_salary': 27215700, 'count': 349})
('Atos', {'total_salary': 24573800, 'count': 313})
('Микрон', {'total_salary': 27246900, 'count': 347})
('Гринатом', {'total_salary': 22899700, 'count': 291})
('ХайТэк', {'total_salary': 26530200, 'count': 336})
('Сател', {'total_salary': 26060100, 'count': 329})
('Лига Цифровой Экономики', {'total_salary': 27095200, 'count': 342})
('TerraLink', {'total_salary': 27271700, 'count': 341})
('АйТеко', {'total_salary': 25771700, 'count': 322})
('Itransition', {'total_salary': 26908400, 'count': 336})
('Эвотор', {'total_salary': 26689300, 'count': 333})
('Angara', {'total_salary': 24971800, 'count': 311})
('OCS Distribution', {'total_salary': 27508100, 'count': 342})
('Hikvision', {'total_salary': 26626700, 'count': 331})
('Тринити', {'total_salary': 27397100, 'count': 339})
(

In [ ]:
import csv
d = {}
with open('salary_data.csv', encoding='utf-8') as file:
    rows = list(csv.reader(file, delimiter=';'))
    for key, value in rows[1:]:
        d[key] = d.get(key, []) + [int(value)]

    d_sort = sorted(d, key=lambda x: (sum(d[x]) / len(d[x]), x)) # упорядочивает список сначала по средней зарплате, затем лексикографически.
    print(*d_sort, sep='\n')

Сортировка по столбцу
Вам доступен файл deniro.csv, каждый столбец которого содержит либо только числа, либо строковые значения:

Machete,2010,72
Marvin's Room,1996,80
Raging Bull,1980,97
...
Напишите программу, которая сортирует содержимое данного файла по указанному столбцу. Причем данные должны быть отсортированы в порядке возрастания чисел, если столбец содержит числа, и в лексикографическом порядке слов, если столбец содержит слова.

Формат входных данных
На вход программе подается натуральное число — номер столбца файла deniro.csv.

Формат выходных данных
Программа должна отсортировать содержимое файла deniro.csv по введенному столбцу и вывести полученный результат в исходном формате.

Примечание 1. Нумерация столбцов начинается с единицы.

Примечание 2. Например, если бы файл deniro.csv имел вид:

red,4
blue,3
green,28
purple,1
и его требовалось отсортировать по второму столбцу (в порядке возрастания чисел), то программа должна была бы вывести:

purple,1
blue,3
red,4
green,28
Примечание 3. Если две какие-либо строки имеют одинаковые значения в столбцах, то следует сохранить их исходный порядок следования.

Примечание 4. Разделителем в файле deniro.csv является запятая, при этом кавычки не используются.

In [151]:
import csv

number = int(input()) - 1

with open('deniro.csv', encoding='utf-8-sig') as file:
    rows = csv.reader(file)
    if number:
        rows = sorted(rows, key=lambda x:(int(x[number])))
    else:
        rows = sorted(rows, key=lambda x:(x[number]))
    for row in rows:
        print(*row, sep=',')


15 Minutes,2001,33
A Bronx Tale,1993,96
Analyze That,2002,27
Analyze This,1999,69
Angel Heart,1987,78
Arthur and the Invisibles,2007,21
Awakenings,1990,88
Backdraft,1991,71
Bang the Drum Slowly,1973,88
Being Flynn,2012,51
Bloody Mama,1970,17
Born to Win,1971,40
Brazil,1985,98
Cape Fear,1991,76
Captain Shakespeare,2007,76
Casino,1995,80
City by the Sea,2002,48
Cop Land,1997,72
Dear America: Letters Home From Vietnam,1987,100
Dirty Grandpa,2016,11
Everybody's Fine,2009,46
Falling in Love,1984,60
Flawless,1999,43
Godsend,2003,4
Goodfellas,1990,96
Great Expectations,1998,38
Greetings,1968,86
Grudge Match,2013,29
Guilty by Suspicion,1991,65
Heat,1995,86
Heist,2015,26
Hide and Seek,2005,13
Jackie Brown,1997,87
Jacknife,1989,64
Joy,2015,60
Killer Elite,2011,25
Killing Season,2013,11
Last Vegas,2013,46
Limitless,2011,70
Little Fockers,2010,10
Machete,2010,72
Mad Dog and Glory,1993,78
Marvin's Room,1996,80
Mary Shelley's Frankenstein,1994,39
Mean Streets,1973,98
Meet the Fockers,2004,38
Meet th

In [ ]:
import csv

i = int(input()) - 1

with open('deniro.csv', 'r', encoding='utf-8') as file:
    data = list(csv.reader(file))

data.sort(key=lambda x: int(x[i]) if x[i].isdigit() else x[i])
for lst in data:
    print(*lst, sep=',')

In [152]:
import csv, sys
n = int(input())-1
with open('deniro.csv') as f:
    csv.writer(sys.stdout).writerows(sorted(csv.reader(f), key=lambda i: int(i[n]) if i[n].isdigit() else i[n]))

15 Minutes,2001,33
A Bronx Tale,1993,96
Analyze That,2002,27
Analyze This,1999,69
Angel Heart,1987,78
Arthur and the Invisibles,2007,21
Awakenings,1990,88
Backdraft,1991,71
Bang the Drum Slowly,1973,88
Being Flynn,2012,51
Bloody Mama,1970,17
Born to Win,1971,40
Brazil,1985,98
Cape Fear,1991,76
Captain Shakespeare,2007,76
Casino,1995,80
City by the Sea,2002,48
Cop Land,1997,72
Dear America: Letters Home From Vietnam,1987,100
Dirty Grandpa,2016,11
Everybody's Fine,2009,46
Falling in Love,1984,60
Flawless,1999,43
Godsend,2003,4
Goodfellas,1990,96
Great Expectations,1998,38
Greetings,1968,86
Grudge Match,2013,29
Guilty by Suspicion,1991,65
Heat,1995,86
Heist,2015,26
Hide and Seek,2005,13
Jackie Brown,1997,87
Jacknife,1989,64
Joy,2015,60
Killer Elite,2011,25
Killing Season,2013,11
Last Vegas,2013,46
Limitless,2011,70
Little Fockers,2010,10
Machete,2010,72
Mad Dog and Glory,1993,78
Marvin's Room,1996,80
Mary Shelley's Frankenstein,1994,39
Mean Streets,1973,98
Meet the Fockers,2004,38
Meet th

Функция csv_columns()
Реализуйте функцию csv_columns(), которая принимает один аргумент:

filename — название csv файла, например, data.csv
Функция должна возвращать словарь, в котором ключом является название столбца файла filename, а значением — список элементов этого столбца.

Примечание 1. Гарантируется, что в передаваемом в функцию файле разделителем является запятая, при этом кавычки не используются.

Примечание 2. Гарантируется, что у передаваемого в функцию файла первая строка содержит названия столбцов.

Примечание 3. Например, если бы файл exam.csv имел вид:

name,grade
Timur,5
Arthur,4
Anri,5
то следующий вызов функции csv_columns():

csv_columns('exam.csv')
должен был бы вернуть:

{'name': ['Timur', 'Arthur', 'Anri'], 'grade': ['5', '4', '5']}
Примечание 4. Ключи в словаре, а также элементы в списках должны располагаться в своем исходном порядке.

Примечание 5. В тестирующую систему сдайте программу, содержащую только необходимую функцию csv_columns(), но не код, вызывающий ее.

In [194]:
import csv

def csv_columns(filename: str):
    with open(filename, encoding='utf-8') as file:
        rows = csv.DictReader(file)
        # for row in rows:
        #     print(row)
        
        new_dic = {}
        for row in rows:
            for key, value in row.items():
                new_dic[key] = new_dic.get(key, [])  # Используем get для получения списка или инициализируем его
                # Метод `get()` используется в Python для получения значения по ключу из словаря. Если ключ не существует в словаре, метод `get()` возвращает значение по умолчанию, которое вы можете задать. Это делает его более безопасным способом доступа к элементам словаря, так как он предотвращает возникновение ошибки `KeyError`, которая возникает, если вы пытаетесь обратиться к ключу, которого нет в словаре.
                new_dic[key].append(value)
                # Метод `append()` добавляет элемент в конец списка. В данном случае, он добавляет `value` в список, соответствующий ключу `key`.
            
        return new_dic
csv_columns('exam.csv')

{'movie': ['Machete',
  "Marvin's Room",
  'Raging Bull',
  "This Boy's Life",
  'Silver Linings Playbook',
  'Taxi Driver',
  'Jackie Brown',
  'Shark Tale',
  'Bang the Drum Slowly',
  'Analyze That',
  'Meet the Parents',
  'Wag the Dog',
  'The Big Wedding',
  'Night and the City',
  'Backdraft',
  'The Untouchables',
  'Cop Land',
  'Thunderheart',
  'Being Flynn',
  "We're No Angels",
  'Limitless',
  'The Bag Man',
  'The Good Shepherd',
  'Jacknife',
  'Righteous Kill',
  'Mad Dog and Glory',
  'Brazil',
  "Mary Shelley's Frankenstein",
  'Stone',
  'Killer Elite',
  'A Bronx Tale',
  'Falling in Love',
  'The Adventures of Rocky & Bullwinkle',
  'Red Lights',
  'The Score',
  "New Year's Eve",
  'Ronin',
  'Midnight Run',
  'Last Vegas',
  'Born to Win',
  'Angel Heart',
  'City by the Sea',
  'Cape Fear',
  "Everybody's Fine",
  'Goodfellas',
  '15 Minutes',
  'Mistress',
  'Hide and Seek',
  'The Intern',
  'Awakenings',
  'Joy',
  'Mean Streets',
  'The Deer Hunter',
  'Gre

In [ ]:
import csv

def csv_columns(filename):

    with open(filename, encoding="utf-8") as file_in:
        rows = list(csv.reader(file_in))
        return {key: value for key, *value in zip(*rows)}

Популярные домены
Вам доступен файл data.csv, который содержит неповторяющиеся данные о пользователях некоторого ресурса. В первом столбце записано имя пользователя, во втором — фамилия, в третьем — адрес электронной почты:

first_name,surname,email
John,Wilson,johnwilson@outlook.com
Mary,Wilson,marywilson@list.ru
...
Напишите программу, которая создает файл domain_usage.csv, имеющий следующее содержание:

domain,count
rambler.ru,24
iCloud.com,29
...
где в первом столбце записано название почтового домена, а во втором — количество пользователей, использующих данный домен. Домены в файле должны быть расположены в порядке возрастания количества их использований, при совпадении количества использований — в лексикографическом порядке.

Примечание 1. Разделителем в файле data.csv является запятая, при этом кавычки не используются.

Примечание 3. Начальная часть файла domain_usage.csv выглядит так:

domain,count
rambler.ru,24
iCloud.com,29
...
Примечание 4. При открытии файла используйте явное указание кодировки UTF-8.

In [270]:
import csv

with open('data.csv', encoding='utf-8-sig') as file:
    table = [r.split(',')[2] for r in file.read().splitlines()]
    del table[0]
    lst = [i.split('@')[1] for i in table]
    print(lst)

    dic ={}
    for domain in lst:
        if domain not in dic:
            dic[domain] = 1
        else:
            dic[domain] += 1        
        
    # print(dic)
    
    result = [{'domain': k, 'count': v} for k, v in dic.items()]
    result = sorted(result, key=lambda x:(x['count'],x['domain']))
    print(*result)
    print(result[1]['domain'])
    print(result[1]['count'])

with open('domain_usage.csv', 'w', encoding='utf-8', newline='') as file2:
    writer = csv.writer(file2)
    writer.writerow(['domain', 'count'])
    for item in result:
        l = [item['domain'], item['count']]
        writer.writerow(l)

['outlook.com', 'list.ru', 'yahoo.com', 'bk.ru', 'mail.ru', 'bk.ru', 'rambler.ru', 'iCloud.com', 'mail.ru', 'mail.ru', 'mail.ru', 'list.ru', 'rambler.ru', 'rambler.ru', 'internet.ru', 'inbox.ru', 'yandex.ru', 'gmail.com', 'yandex.ru', 'internet.ru', 'internet.ru', 'yandex.ru', 'rambler.ru', 'rambler.ru', 'mail.ru', 'yahoo.com', 'internet.ru', 'gmail.com', 'yahoo.com', 'gmail.com', 'gmail.com', 'aol.com', 'internet.ru', 'yandex.ru', 'yahoo.com', 'inbox.ru', 'gmail.com', 'outlook.com', 'gmail.com', 'yandex.ru', 'yandex.ru', 'internet.ru', 'mail.ru', 'gmail.com', 'mail.ru', 'yahoo.com', 'list.ru', 'aol.com', 'inbox.ru', 'iCloud.com', 'bk.ru', 'list.ru', 'gmail.com', 'iCloud.com', 'gmail.com', 'outlook.com', 'mail.ru', 'outlook.com', 'inbox.ru', 'rambler.ru', 'yahoo.com', 'gmail.com', 'inbox.ru', 'bk.ru', 'yahoo.com', 'yandex.ru', 'outlook.com', 'gmail.com', 'inbox.ru', 'outlook.com', 'list.ru', 'outlook.com', 'yandex.ru', 'list.ru', 'list.ru', 'yandex.ru', 'list.ru', 'gmail.com', 'iCloud.

In [ ]:
import csv

domens = dict()

with open('data.csv', encoding='utf-8') as f:
    for *n,d in csv.reader(f):
        key = d.split('@')[-1]
        domens[key] = domens.get(key, 0) + 1
        
del domens['email']

with open('domain_usage.csv', 'w', encoding='utf-8', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(['domain','count'])
    for row in sorted(domens.items(), key=lambda x: (x[1], x[0])):
        writer.writerow(row)

Wi-Fi Москвы
Вам доступен файл wifi.csv, который содержит данные о городском Wi-Fi Москвы. В первом столбце записано название округа, во втором — название района, в третьем — адрес, в четвертом — количество точек доступа по этому адресу:

adm_area;district;location;number_of_access_points
Центральный административный округ;район Якиманка;город Москва, улица Серафимовича, дом 5/16;5
Центральный административный округ;район Якиманка;город Москва, Болотная набережная, дом 11, строение 1;2
...
Напишите программу, которая определяет количество точек доступа в каждом районе Москвы и выводит названия всех районов, для каждого указывая соответствующее количество точек доступа, каждое на отдельной строке, в следующем формате:

<название района>: <количество точек доступа>
Названия районов должны быть расположены в порядке убывания количества точек доступа, при совпадении количества точек доступа — в лексикографическом порядке.

Примечание 1. Разделителем в файле wifi.csv является точка с запятой, при этом кавычки не используются.

Примечание 2. При сортировке названия районов должны быть использованы именно в том виде, в котором они указаны в исходном файле. Выполнять какие-либо дополнительные преобразования не нужно.

Примечание 4. Начальная часть ответа выглядит так:

Тверской район: 480
район Хамовники: 386
Пресненский район: 349
...
Примечание 5. При открытии файла используйте явное указание кодировки UTF-8.

In [302]:
import csv
from collections import defaultdict

'''one way'''
dic_first = {}
with open('wifi.csv', encoding='utf-8') as f:
    rows = csv.DictReader(f, delimiter=';')
    for row in rows:
        # print(row)
        district = row['district']
        number_of_access_points = int(row['number_of_access_points'])
        if district not in dic_first.keys():
            dic_first[district] = number_of_access_points
        else:
            dic_first[district] += number_of_access_points
    dic_first = dict(sorted(dic_first.items(), key=lambda x:(-x[1], x[0]), reverse=False))
    # print(dic_first)
    # for key, value in dic_first.items():
    #     print(f'{key}: {value}', sep='')

'''second way'''
dic_second = {}
with open('wifi.csv', encoding='utf-8') as f:
    rows = list(csv.DictReader(f, delimiter=';'))
    dic_second = {
    district: sum(int(row['number_of_access_points']) for row in rows if row['district'] == district)
    for district in {row['district'] for row in rows}  # Уникальные районы
}
    dic_second = dict(sorted(dic_second.items(), key=lambda x:(-x[1], x[0]), reverse=False))
    # print(dic_second)
    # for key, value in dic_second.items():
    #     print(f'{key}: {value}', sep='')
      
'''third way'''
dic_default = defaultdict(int)
with open('wifi.csv', encoding='utf-8') as f:
    rows = csv.DictReader(f, delimiter=';')
    for row in rows:
        dic_default[row['district']] += int(row['number_of_access_points'])
    dic_default = dict(sorted(dic_default.items(), key=lambda x:(-x[1], x[0]), reverse=False))
    # print(dic_default)
    for key, value in dic_default.items():
        print(f'{key}: {value}', sep='')

Тверской район: 480
район Хамовники: 386
Пресненский район: 349
Басманный район: 343
район Замоскворечье: 308
Мещанский район: 228
Таганский район: 212
район Якиманка: 201
район Арбат: 184
Красносельский район: 142
район Дорогомилово: 92
Донской район: 51
Даниловский район: 47
район Беговой: 35
район Марьина Роща: 18
Южнопортовый район: 6
Гагаринский район: 4
Хорошёвский район: 4
район Марьино: 4
район Проспект Вернадского: 2
район Лефортово: 1


In [303]:
import csv

with open('wifi.csv', encoding='UTF-8') as f:
    d = {}
    for r in [*csv.reader(f, delimiter=';')][1:]:
        d[r[1]] = d.get(r[1], 0) + int(r[3])

for i in sorted(d, key=lambda x:(-d[x], x)):
    print(f'{i}: {d[i]}')

Тверской район: 480
район Хамовники: 386
Пресненский район: 349
Басманный район: 343
район Замоскворечье: 308
Мещанский район: 228
Таганский район: 212
район Якиманка: 201
район Арбат: 184
Красносельский район: 142
район Дорогомилово: 92
Донской район: 51
Даниловский район: 47
район Беговой: 35
район Марьина Роща: 18
Южнопортовый район: 6
Гагаринский район: 4
Хорошёвский район: 4
район Марьино: 4
район Проспект Вернадского: 2
район Лефортово: 1


Последний день на Титанике
Вам доступен файл titanic.csv, который содержит данные о пассажирах, присутствовавших на борту парохода Титаник. В первом столбце указана единица, если пассажир выжил, и ноль в противном случае, во втором столбце записано полное имя пассажира, в третьем — пол, в четвертом — возраст:

survived;name;sex;age
0;Mr. Owen Harris Braund;male;22
1;Mrs. John Bradley (Florence Briggs Thayer) Cumings;female;38
...
Напишите программу, которая выводит имена выживших пассажиров, которым было менее 
18
18 лет, каждое на отдельной строке. Причем сначала должны быть расположены имена всех пассажиров мужского пола, а затем — женского, имена же непосредственно в мужском и женском списках должны быть расположены в своем исходном порядке.

Примечание 1. Разделителем в файле titanic.csv является точка с запятой, при этом кавычки не используются.

Примечание 3. Часть ответа выглядит так:

Master. Gerios Moubarek
Master. Alden Gates Caldwell
...
Master. Harold Theodor Johnson
Mrs. Nicholas (Adele Achem) Nasser
Miss. Marguerite Rut Sandstrom
...
Примечание 4. При открытии файла используйте явное указание кодировки UTF-8.

In [314]:
import csv

with open('titanic.csv', encoding='utf-8') as f:
    rows = csv.DictReader(f, delimiter=';')
    filtered_rows = filter(lambda x:x['survived']=='1' and float(x['age'])<18, rows)
    sorted_rows = sorted(filtered_rows, key=lambda x:len(x['sex']))
    for row in sorted_rows:
        print(row['name'])

Master. Gerios Moubarek
Master. Alden Gates Caldwell
Master. Elias Nicola-Yarred
Master. Frank John William Goldsmith
Master. Richard F Becker
Master. Michel M Navratil
Mr. Victor Francis Sunderland
Master. Edvin Rojj Felix Asplund
Master. Hudson Trevor Allison
Master. Edmond Roger Navratil
Master. William Loch Coutts
Master. William Rowe Richards
Master. Washington Dodge
Master. Eden Leslie Coutts
Master. John Morgan Jr Davies
Mr. John Borland Jr Thayer
Master. Halim Gonios Moubarek
Master. Meier Moor
Master. Viljo Hamalainen
Master. Bertram Vere Dean
Master. William Thornton II Carter
Master. Assad Alexander Thomas
Master. Andre Mallet
Master. George Sibley Richards
Master. Harold Theodor Johnson
Mrs. Nicholas (Adele Achem) Nasser
Miss. Marguerite Rut Sandstrom
Miss. Anna McGowan
Miss. Jamila Nicola-Yarred
Miss. Simonne Marie Anne Andree Laroche
Miss. Constance Mirium West
Miss. Erna Alexandra Andersson
Miss. Bertha Ilett
Miss. Anna Peter
Miss. Katherine Gilnagh
Miss. Eleanor Ileen J

In [316]:
import pandas as pd

df = pd.read_csv('titanic.csv', delimiter=';')
# print(df)
df_new = df.loc[(df.age < 18) & (df.survived == 1)]

print(*df_new.loc[df.sex == 'male'].name, sep='\n')
print(*df_new.loc[df.sex == 'female'].name, sep='\n')

     survived                                               name     sex   age
0           0                             Mr. Owen Harris Braund    male  22.0
1           1  Mrs. John Bradley (Florence Briggs Thayer) Cum...  female  38.0
2           1                              Miss. Laina Heikkinen  female  26.0
3           1        Mrs. Jacques Heath (Lily May Peel) Futrelle  female  35.0
4           0                            Mr. William Henry Allen    male  35.0
..        ...                                                ...     ...   ...
882         0                               Rev. Juozas Montvila    male  27.0
883         1                        Miss. Margaret Edith Graham  female  19.0
884         0                     Miss. Catherine Helen Johnston  female   7.0
885         1                               Mr. Karl Howell Behr    male  26.0
886         0                                 Mr. Patrick Dooley    male  32.0

[887 rows x 4 columns]


In [23]:
import pandas as pd

df = pd.read_csv('titanic.csv', delimiter=';')
df = df[(df.survived == 1) & (df.age < 18)]
print(*df[df.sex == 'male'].name, sep='\n')
print(*df[df.sex == 'female'].name, sep='\n')

Master. Gerios Moubarek
Master. Alden Gates Caldwell
Master. Elias Nicola-Yarred
Master. Frank John William Goldsmith
Master. Richard F Becker
Master. Michel M Navratil
Mr. Victor Francis Sunderland
Master. Edvin Rojj Felix Asplund
Master. Hudson Trevor Allison
Master. Edmond Roger Navratil
Master. William Loch Coutts
Master. William Rowe Richards
Master. Washington Dodge
Master. Eden Leslie Coutts
Master. John Morgan Jr Davies
Mr. John Borland Jr Thayer
Master. Halim Gonios Moubarek
Master. Meier Moor
Master. Viljo Hamalainen
Master. Bertram Vere Dean
Master. William Thornton II Carter
Master. Assad Alexander Thomas
Master. Andre Mallet
Master. George Sibley Richards
Master. Harold Theodor Johnson
Mrs. Nicholas (Adele Achem) Nasser
Miss. Marguerite Rut Sandstrom
Miss. Anna McGowan
Miss. Jamila Nicola-Yarred
Miss. Simonne Marie Anne Andree Laroche
Miss. Constance Mirium West
Miss. Erna Alexandra Andersson
Miss. Bertha Ilett
Miss. Anna Peter
Miss. Katherine Gilnagh
Miss. Eleanor Ileen J

Лог-файл
Вам доступен файл name_log.csv, в котором находятся логи изменения имени пользователя. В первом столбце записано измененное имя пользователя, во втором — адрес электронной почты, в третьем — дата и время изменения. При этом email пользователь менять не может, только имя:

username,email,dtime
rare_charles6,charlesthompson@inbox.ru,15/11/2021 08:15
busy_patricia5,patriciasmith@bk.ru,07/11/2021 08:07
...
Напишите программу, которая отбирает из файла name_log.csv только самые свежие записи для каждого пользователя и записывает их в файл new_name_log.csv. В файле new_name_log.csv первой строкой должны быть заголовки столбцов такие же, как в файле name_log.csv. Логи в итоговом файле должны быть расположены в лексикографическом порядке названий электронных ящиков пользователей.

Примечание 1. Для части пользователей в исходном файле запись только одна, и тогда в итоговый файл следует записать только ее, для некоторых пользователей есть несколько записей с разными именами.

Например, пользователь с электронной почтой c3po@gmail.com несколько раз менял имя:

C=3PO,c3po@gmail.com,16/11/2021 17:10
C3PO,c3po@gmail.com,16/11/2021 17:15
C-3PO,c3po@gmail.com,16/11/2021 17:24
Из этих трех записей в итоговый файл должна быть записана только одна — самая свежая:

C-3PO,c3po@gmail.com,16/11/2021 17:24
Примечание 2. Разделителем в файле name_log.csv является запятая, при этом кавычки не используются.

Примечание 4. Начальная часть файла new_name_log.csv выглядит так:

username,email,dtime
angry-barbara2,barbaraanderson@bk.ru,17/11/2021 01:17
dead-barbara6,barbarabrown@rambler.ru,27/11/2021 08:27
busy_barbara7,barbaradavis@aol.com,24/11/2021 08:24
...
Примечание 5. При открытии файла используйте явное указание кодировки UTF-8.

In [73]:
import csv
from itertools import islice
from datetime import datetime

list_of_dic = []
with open('name_log.csv', encoding='utf-8') as f:
    rows = csv.DictReader(f)
    # получаю строки для понимания структуры
    # for row in islice(rows, 2):
    for row in rows:
        # print(row)
        
        # Преобразуем строку в объект datetime
        row['dtime'] = datetime.strptime(row['dtime'], '%d/%m/%Y %H:%M')
        # print(row)
        
        list_of_dic.append(row)
    
# print(list_of_dic)

sorted_list = sorted(list_of_dic, key=lambda x:x['dtime'], reverse=True)
sorted_list = sorted(sorted_list, key=lambda x:x['email'])
# print(sorted_list)

temp = []
final = []
for row in sorted_list:
    if row['email'] not in temp:
        temp.append(row['email'])
        final.append((row['username'], row['email'], datetime.strftime(row['dtime'], '%d/%m/%Y %H:%M')))
        
# print(final)
title = list((sorted_list[0]).keys())
# print(title)

with open('new_name_log.csv', 'w', encoding='utf-8', newline='') as f:
    writer = csv.writer(f)
    writer.writerow(title)
    for row in final:
        writer.writerow(row)

In [ ]:
import csv
from datetime import datetime

with open('name_log.csv', encoding='UTF-8') as f:
	header, *rows = csv.reader(f)

d = {i[1]:i for i in sorted(rows, key=lambda x: datetime.strptime(x[2], '%d/%m/%Y %H:%M'))}

with open('new_name_log.csv', 'w', encoding='UTF-8', newline='') as f:
	w = csv.writer(f)
	w.writerow(header)
	w.writerows(sorted(d.values(), key=lambda x: x[1]))

Проще, чем кажется 🌶️
Рассмотрим следующий текстовый фрагмент:

ball,color,purple
ball,size,4
ball,notes,it's round
cup,color,blue
cup,size,1
cup,notes,none
Каждая строка этого фрагмента содержит три значения через запятую: имя объекта, свойство этого объекта, значение свойства. Например, в первой строке указан объект ball, имеющий свойство color, значение которого равно purple. Также у объекта ball есть свойства size и notes, имеющие значения 4 и it's round соответственно. Помимо объекта ball имеется объект cup, имеющий те же свойства и в том же количестве. Дадим этим объектам общее название object и сгруппируем строки данного текстового фрагмента по первому столбцу:

object,color,size,notes
ball,purple,4,it's round
cup,blue,1,none
Мы получили запись в привычном CSV формате, в котором в первом столбце указывается имя объекта, а в последующих — значения соответствующих свойств этого объекта.

Реализуйте функцию condense_csv(), которая принимает два аргумента в следующем формате:

filename — название csv файла, например, data.csv; формат содержимого файла аналогичен формату текстового фрагмента, рассмотренного в условии задачи: каждая строка файла содержит три значения через запятую, а именно имя объекта, свойство этого объекта, значение свойства; все объекты имеют равные свойства и в равных количествах
id_name — общее название для объектов
Функция должна привести содержимое файла в привычный CSV формат, сгруппировав строки по первому столбцу и назвав первый столбец id_name. Полученный результат функция должна записать в файл condensed.csv.

Примечание 1. Например, если бы файл data.csv имел следующий вид:

01,Title,Ran So Hard the Sun Went Down
02,Title,Honky Tonk Heroes (Like Me)
то вызов функции condense_csv():

condense_csv('data.csv', id_name='ID')
должен был бы создать файл condensed.csv со следующим содержанием:

ID,Title
01,Ran So Hard the Sun Went Down
02,Honky Tonk Heroes (Like Me)
Примечание 2. Гарантируется, что в передаваемом в функцию csv файле разделителем является запятая, при этом кавычки не используются.

Примечание 3. При открытии файла используйте явное указание кодировки UTF-8.

Примечание 4. В тестирующую систему сдайте программу, содержащую только необходимую функцию condense_csv(), но не код, вызывающий ее.

In [118]:
import csv


# берём текст
text = '''01,Artist,Otis Taylor
01,Title,Ran So Hard the Sun Went Down
01,Time,3:52
02,Artist,Waylon Jennings
02,Title,Honky Tonk Heroes (Like Me)
02,Time,3:29
03,Artist,David Allan Coe
03,Title,Willie Waylon And Me
03,Time,3:26'''

# записываем текст в файл
with open('data1.csv', 'w', encoding='utf-8') as file:
    file.write(text)

def condense_csv(filename: str, id_name: str):
    header = []
    with open(filename, encoding='utf-8') as file:
        rows = csv.reader(file)
        for row in rows:
            if row[1] not in header:
                header.append(row[1])
    header.insert(0, id_name)
    # print(header)
    # print()
    
    lst = []
    dic = {}
    with open(filename, encoding='utf-8') as file:
        rows = csv.reader(file)
        for row in rows:
            lst.append(row)
            if row[0] not in dic.keys():
                dic[row[0]] = {row[1]:row[2]}
            else:
                # Обновим существующий словарь.
                dic[row[0]].update({row[1]: row[2]})
    # print(lst)
    # print()
    # print(dic)
    # print()
    
    final_list = []
    for key, value in dic.items():
        # print(key, *value.values(), sep=',')
        s = (key, *value.values())
        final_list.append(s)
    
    with open('condensed.csv', 'w', encoding='utf-8', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(header)
        for row in final_list:         
            writer.writerow(row)

# создаём новый файл condensed.csv
condense_csv('data1.csv', id_name='Position')

# выводим на печать получившийся текст из созданного файла
with open('condensed.csv', encoding='utf-8') as file:
    print(file.read().strip())

Position,Artist,Title,Time
01,Otis Taylor,Ran So Hard the Sun Went Down,3:52
02,Waylon Jennings,Honky Tonk Heroes (Like Me),3:29
03,David Allan Coe,Willie Waylon And Me,3:26


In [ ]:
import csv

def condense_csv(filename, id_name):
    with open(filename, encoding='utf-8') as file:
        objects = {}
        for obj, attr, value in csv.reader(file):
            if obj not in objects:
                objects[obj] = {id_name: obj}
            objects[obj][attr] = value
    
    with open('condensed.csv', 'w', encoding='utf-8') as file:
        writer = csv.DictWriter(file, fieldnames=objects[obj])
        writer.writeheader()
        writer.writerows(objects.values())

Возрастание классов 🌶️
Вам доступен файл student_counts.csv, который содержит данные о количестве учеников в некотором учебном заведении за период 
2000
2000 — 
2021
2021 г. В первом столбце записан год, в последующих столбцах записан класс и количество учеников в данном классе в этом году:

year,5-Б,3-Б,8-А,2-Г,7-Б,1-Б,3-Г,3-А,2-В,6-Б,6-А,8-Б,8-Г,11-А,2-А,7-А,5-А,2-Б,10-А,11-Б,8-В,4-А,7-В,3-В,1-А,9-А,11-В
2000,19,15,18,29,19,17,26,29,28,30,26,27,27,22,29,19,27,20,16,18,15,27,19,29,22,20,23
2001,21,30,22,19,26,20,24,27,20,30,24,30,29,21,20,19,29,27,23,25,30,30,23,22,22,18,22
...
Напишите программу, которая записывает данную таблицу в файл sorted_student_counts.csv, располагая все столбцы в порядке возрастания классов, при совпадении классов — в порядке возрастания букв.

Примечание 1. Каждый класс содержит номер и букву и записывается в следующем формате:

<номер класса>-<буква класса>
Примечание 2. Разделителем в файле student_counts.csv является запятая, при этом кавычки не используются.

Примечание 4. Начальная часть файла sorted_student_counts.csv выглядит так:

year,1-А,1-Б,2-А,2-Б,...
2000,22,17,29,20,...
2001,22,20,20,27,...
...
Примечание 5. При открытии файла используйте явное указание кодировки UTF-8.

In [1]:
import csv

with open('student_counts.csv', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    # Метод list(reader) считывает все данные из CSV-файла и переводит их в список, но после этого объект reader больше не имеет данных для чтения.
    records = list(reader)
    header = reader.fieldnames
    # print(header)
    # print()
    # print(records)
    # print()
    columns = [i for i in records[0]]
    # print(columns)
    # print(header==columns)
    # print()
    # Мысль такая: в методе записи DictWriter - Обязательный аргумент - fieldnames (название столбцов), в словарях ключом является - название столбца. Т.е. если в DictWriter - аргумент fieldnames = отсортированному списку названий столбцов и загрузить в writerows(тот же словарь из DictReader) - то получим отсортированную по наименованию столбцов таблицу.
    header = [reader.fieldnames[0]] + sorted(reader.fieldnames[1:], key=lambda x: (int(x.split('-')[0]), x[2]))
    # print(header)
    # print()

with open('sorted_student_counts.csv', 'w', encoding='utf-8', newline='') as file:
    writer = csv.DictWriter(file, fieldnames=header)
    writer.writeheader()
    for row in records:
        writer.writerow(row)

['year', '5-Б', '3-Б', '8-А', '2-Г', '7-Б', '1-Б', '3-Г', '3-А', '2-В', '6-Б', '6-А', '8-Б', '8-Г', '11-А', '2-А', '7-А', '5-А', '2-Б', '10-А', '11-Б', '8-В', '4-А', '7-В', '3-В', '1-А', '9-А', '11-В']


In [3]:
import csv

def key_func(grade):
    number, letter = grade.split('-')
    return int(number), letter

with open('student_counts.csv', encoding='utf-8') as file:
    reader = csv.DictReader(file)
    columns = ['year'] + sorted(reader.fieldnames[1:], key=key_func)
    rows = list(reader)

with open('sorted_student_counts.csv', 'w', encoding='utf-8') as file:
    writer = csv.DictWriter(file, fieldnames=columns)
    writer.writeheader()
    writer.writerows(rows)

Голодный студент 🌶️
Дима очень хочет поесть, но денег у него мало. Помогите ему определить самый дешевый продукт, а также магазин, в котором он продается. Вам доступен файл prices.csv, который содержит информацию о ценах продуктов в различных магазинах. В первом столбце записано название магазина, а в последующих — цена на соответствующий товар в этом магазине:

Магазин;Творог;Гречка;Рис;Бородинский хлеб;Яблоки;Пельмени;Овсяное печенье;Спагетти;Печеная фасоль;Мороженое;Фарш;Вареники;Картофель;Батончик
Пятерочка;69;133;129;83;141;90;72;123;149;89;88;106;54;84
Магнит;102;87;95;75;109;112;97;82;101;134;69;61;141;79
...
Напишите программу, которая определяет и выводит самый дешевый продукт и название магазина, в котором он продается, в следующем формате:

<название продукта>: <название магазина>
Если имеется несколько самых дешевых товаров, то следует вывести тот товар, чье название меньше в лексикографическом сравнении. Если один товар продается в нескольких магазинах по одной минимальной цене, то следует вывести тот магазин, чье название меньше в лексикографическом сравнении.

Примечание 1. Разделителем в файле prices.csv является точка с запятой, при этом кавычки не используются.

Примечание 3. Пример вывода:

Клубничный йогурт: ВкусВилл
Примечание 4. При открытии файла используйте явное указание кодировки UTF-8.

In [35]:
import csv

with open('prices.csv', encoding='utf-8') as file:
    reader = csv.DictReader(file, delimiter=';')
    # header = reader.fieldnames
    # print(header)
    # print(len(header))
    # print()
    records = list(reader)
    # print(records)
    # print()
    
    new_dic = {}
    for dic in records:
        store = dic['Магазин']  # Название магазина
        keys = list(dic.keys())
        # print(keys)
        # Находим ключ с минимальным значением, приводя значения к типу int
        min_product_name = min(keys[1:], key=lambda k: int(dic[k]))
        # print(min_product_name)
        min_product_value = dic[min_product_name]
        # print(min_product_value)
        # if store not in new_dic:
        # new_dic[min_product_name] = (store, int(min_product_value))
        new_dic[store] = (min_product_name, int(min_product_value))

            
    # print()
    # print(new_dic)
    # print()
    final_dic = sorted(new_dic.items(), key= lambda i: (i[1][1], i[1][0]))
    # print(final_dic[0])
    print(f'{final_dic[0][1][0]}: {final_dic[0][0]}')

Вареники: Дикси
